lib/elixir/lib/enum.ex

```diff
@@ -2734,11 +2734,15 @@ defmodule Enum do
       [{1, :a}, {2, :b}, {3, :c}]
 
   """
-  @spec zip([t]) :: t
+  @spec zip([t] | %Stream{}) :: t
 
   def zip([]), do: []
 
-  def zip(enumerables) when is_list(enumerables) do
+  def zip(%Stream{} = enumerables), do: do_zip(enumerables)
+
+  def zip(enumerables) when is_list(enumerables), do: do_zip(enumerables)
+
+  defp do_zip(enumerables) do
```

Comment:
This would be much cleaner if we can have a guard that checks that `enumerable` is a list or a `%Stream{}` struct but I'm not sure how to do that.